### ISSUE TITLE
DUNST_TIMESTAMP variable meaning?

### ISSUE THREAD

**Issue by tagwint** [dunst-project/dunst#1019]
### Installation info

- Version: `Dunst - A customizable and lightweight notification-daemon 1.6.1`
- Install type: `package|manually`
- Window manager / Desktop environment: `i3wm`

<details>
<summary>Minimal dunstrc</summary>

```ini
[global]
    allow_markup = yes
    format = "App:%a\n<b>Summary:</b>\n%s\n<b>Body:</b>\n%b"
    sort = yes
    indicate_hidden = yes
    alignment = left
    bounce_freq = 0
    show_age_threshold = 60
    word_wrap = yes
    ignore_newline = no
    geometry = "300x5-30+20"
    shrink = no
    transparency = 0
    idle_threshold = 120
    monitor = 0
    follow = mouse
    sticky_history = yes
    history_length = 20
    show_indicators = yes
    line_height = 0
    separator_height = 2
    padding = 8
    horizontal_padding = 8
    separator_color = frame
    startup_notification = false
    dmenu = /usr/bin/dmenu -p dunst:
    browser = /usr/bin/google-chrome
    always_run_script = true
    icon_position = left
    max_icon_size=32
    corner_radius=18

[shortcuts]

    close = ctrl+space
    close_all = ctrl+shift+space
    history = ctrl+shift+grave

[urgency_low]
    background = "#C6BEE4"
    foreground = "#6B6B6B"
    timeout = 10

[urgency_normal]
    background = "#332b2e"
    foreground = "#ffffff"
    timeout = 1d

[urgency_critical]
    background = "#900000"
    foreground = "#ffffff"
    timeout = 3d

[logging]
    summary = "*"
    script = /tmp/dunst_log.sh
```
</details>

<details>
<summary>Logging script</summary>

```shell
#!/bin/bash

EDEBUG=/tmp/dunstlog2
echo "SYSTEM_TIMESTAMP ================ $(date +'%s')">>$EDEBUG
echo "DUNST_TIMESTAMP ================= $DUNST_TIMESTAMP)">>$EDEBUG
```
</details>

<details>
<summary>Script output</summary>

```shell
SYSTEM_TIMESTAMP ================ 1642419690
DUNST_TIMESTAMP ================= 515285298
SYSTEM_TIMESTAMP ================ 1642420270
DUNST_TIMESTAMP ================= 515864613
```
</details>

**Expected result:**
Environment variable `DUNST_TIMESTAMP` to contain message timestamp in seconds since 1970-01-01 00:00:00 UTC
 
**Actual result:**
`DUNST_TIMESTAMP` value contain a number of unknown (to me) meaning

**Comment by fwsmit** [comment]
Right now it seems like it's the result of `clock_gettime` in seconds(or milliseconds?). It's a monotonically increasing time with a meaningless origin. It could be used to compare if a notification was sent earlier than another, but you cannot see the absolute time at which it was sent.

Of course this is subject to change, since the current one might not be very useful. Do you have a use case for it?

**Comment by tagwint** [comment]
Yes, the use case is I want to see the timestamp in the message body, not just the age like X hours ago. Since there's no direct way to have that (is there?), I was about to modify the body and by injecting the timestamp and update the message by  `dunstify -r <id>`. Do you think there's a better idea of implementing that?
(thanks for quick response)

**Comment by fwsmit** [comment]
The best way of doing this would be to add a few options to `format` similar to `date(1)`

**Comment by Nickwiz** [comment]
From what I can find `DUNST_TIMESTAMP` is system up-time in milliseconds for when message was sent.

I.e. Uptime starts at 0 on boot. If a message is sent 634 seconds after boot, DUNST_TIMESTAMP will be 634 for that message.

One can for example use `uptime -s`, which gives time for when `uptime` started to tick or
first entry in `/proc/uptime` on Linux gives seconds since boot, with possible fraction.

A (hackish) way to get the time could be something like:

```sh
dunst_ts_to_unix() {
	# Convert DUNST_TIMESTAMP from ms to seconds
	local dunst_ts=$(( $1 / 1000))
	# Get seconds since boot
	local seconds_uptime=$(date +%s -d@$(cut -d' ' -f1 /proc/uptime))
	# Get seconds as UNIX timestamp
	local ts_now=$(date +%s)
	# Print the diff of: NOW - (uptime - dunst-timestamp)
	printf %d $((ts_now - seconds_uptime + dunst_ts))
}

# Use as for example:
date +'%F %T' -d@$(dunst_ts_to_unix "$DUNST_TIMESTAMP")

```

If one want the fractions one can format by for example `%s%3N` when working with the numbers.

**Comment by jinnko** [comment]
FWIW - I took the sample shell script from @Nickwiz and turned it into some python.  In doing so, on my system (arch with lts kernel), the incoming `$DUNST_TIMESTAMP` seems to be in nanoseconds.

```python
def get_uptime():
    with open("/proc/uptime", "r") as f:
        uptime_seconds = float(f.readline().split()[0])
    return uptime_seconds


def dunst_ts_to_time(ts):
    # Convert DUNST_TIMESTAMP from ns to seconds
    dunst_ts = int(ts / 1000 / 1000)
    seconds_uptime = get_uptime()
    ts_now = datetime.now().timestamp()
    # convert dunst ts to unix ts
    ts = int(ts_now) - int(seconds_uptime) + int(dunst_ts)
    return datetime.utcfromtimestamp(ts).strftime("%H:%M")
```

**Comment by spacelama** [comment]
Even that seems implementation specific.  On my desktop, which just happens to be a VM, that just yields a silly result.  The hint may be in the manpage, with CLOCK_MONOTONIC not talking about milliseconds at all.  Perhaps on modern NOHZ kernels that aren't ticking 1000 times per second, the value really is just precisely what it says - a counter that never decrements.

So the question becomes: why is dunst using CLOCK_MONOTONIC at all?  There's plenty of use-cases for specialised precision interval timekeeping, but is this one of them?  Oh no, a leapsecond has happened once every 5 years.  Do I really care that my notification came in 65 seconds ago and not 64 seconds ago?  You could get away fine with using CLOCK_REALTIME, but heck, if you really care about monotonicity, just use CLOCK_TAI -- it's currently ~38 seconds out.  But at least then $DUNST_TIMESTAMP and `dunstctl history | grep -A3 timestamp` bear some resemblance to something useful.

**Comment by thoughtsunificator** [comment]
How about documenting this? Maybe we can save earth and its humans A LOT of time. I don't get how:

```
       history
           Exports the history as JSON.
```

How about trying to explain anything, anything at all?

**Comment by fedang** [comment]
> How about documenting this? Maybe we can save earth and its humans A LOT of time. I don't get how:
> 
> ```
>        history
>            Exports the history as JSON.
> ```
> 
> How about trying to explain anything, anything at all?


You mean the timestamp should be documented? Or what else?
